spec/controllers/products_controller_spec.rb

```diff
@@ -0,0 +1,54 @@
+require 'rails_helper'
+
+RSpec.describe Admin::ProductsController, type: :controller do
+  let!(:user) { User.create(name: 'Itachi Uchiha', email: 'itachi.uchihaa@yahoo.com', password: 'Password1!', role: :admin) }
+  let!(:order) { Order.create(user: user, status: :pending) }
+  let(:product_params) { { title: 'New Product', description: 'Product description', price: 9.99, category: :entree, vegetarian: :vegetarian } }
+
+  describe '#show' do
+    it 'renders show template' do
+      product = Product.create(product_params)
+      get :show, params: { id: product.id }
+      expect(response).to render_template(:show)
+    end
+  end
+
+  describe '#new' do
+    before do
+      allow(controller).to receive(:current_user).and_return(user)
+      allow(controller).to receive(:logged_in?).and_return(true)
+    end
+
+    it 'renders new template' do
+      get :new
+      expect(response).to render_template(:new)
+    end
+  end
+
+  describe '#create' do
```

Comment:
I left a similar comment on Laurentiu's PR (copied below), he used it in a test where he created users, not products, so you will need to update the example with `product` instead of `user`

Reply:
There are multiple places where you do the same action:
```ruby
post :create, params: { user: valid_attributes }
```

So what you could do here, is extract that into a `subject`. Your code will look like:
```ruby
describe 'POST #create' do
  let(:valid_create_attributes) do
    {
      name: 'John Doe',
      email: generate_unique_email('johndoe@yahoo.com'),
      password: 'password123',
      password_confirmation: 'password123'
    }
  end

  subject { post :create, params: { user: valid_attributes } }

  context 'with valid attributes' do
    it 'creates a new user' do
      expect { subject }.to change(User, :count).by(1)
    end

    # all tests where the attributes are valid
  end

  # in the context where attributes are not valid
  context 'with valid attributes' do
    # here you modify, add or remove any attribute you want, to make the final params structure to be invalid
    before { valid_create_attributes.merge!({}) }

    # add all the tests with this format of attributes, where they are invalid
  end
end
```

The `subject` is basically an action, with given params. In certain contexts, you will have a `before` action, that will modify the `valid_attributes` variable. So the subject will end up doing the same action, but with the modified `valid_attributes`. So the action (calling `post :create`) will be the same, but the params can be different